src/sass_types.h

```diff
@@ -0,0 +1,225 @@
+#ifndef SASS_TYPES_H
+#define SASS_TYPES_H
+
+#include <nan.h>
+#include "sass_context_wrapper.h"
+
+
+namespace SassTypes 
+{
+  // This is the interface that all sass values must comply with
+  class Value {
+    public:
+      Value(Sass_Value* v) { this->value = sass_clone_value(v); }
+      virtual ~Value() { sass_delete_value(this->value); }
+      virtual Sass_Value* get_sass_value() =0;
+      virtual Local<Object> get_js_object() =0;
+
+    protected:
+      Sass_Value* value;
+  };
+
+
+  // This is the guru that knows everything about instantiating the right subclass of SassTypes::Value
+  // to wrap a given Sass_Value object.
+  class Factory {
+    public:
+      static void initExports(Handle<Object>);
+      static Value* create(Sass_Value*);
+      static Value* unwrap(Handle<v8::Value>);
+  };
+
+
+  // Include this in any SassTypes::Value subclasses to handle all the heavy lifting of constructing JS
+  // objects and wrapping sass values inside them
+  template <class T>
+  class CoreValue : public Value {
+    public:
+      static char const* get_constructor_name() { return "SassValue"; }
```

Comment:
I did not know about this `constexpr` keyword, looks interesting, thanks for the pointers! You are right we probably don't need a function here. As for virtual inheritance, that was a no-go because the constructor name is used from `get_constructor()` which is static.